internal/codegen/tpl_module.go

```diff
@@ -0,0 +1,207 @@
+	s   *Stencil
+	t   *Template
+	log slogext.Logger
+}
+
+// Export registers a function to allow it to be called by other
+// templates.
+//
+// This is only able to be called in library templates and the
+// function's name must start with a capital letter. Function names are
+// also only eligible to be exported once, if a function is exported
+// twice the second call will be a runtime error.
+//
+// Example:
+//
+//	{{- define "HelloWorld" }}
+//	{{- return (printf "Hello, %s!" .Data) }}
+//	{{- end }}
+//
+//	{{ module.Export "HelloWorld" }}
+func (tm *TplModule) Export(name string) (string, error) {
+	tm.s.moduleCaller.mu.Lock()
+	defer tm.s.moduleCaller.mu.Unlock()
+
+	// We only allow functions to be exported in the first pass.
+	if !tm.s.isFirstPass {
+		return "", nil
+	}
+
+	if !tm.t.Library {
+		return "", fmt.Errorf("only library templates can export functions")
+	}
+
+	if !strings.HasPrefix(name, strings.ToUpper(name[:1])) {
+		return "", fmt.Errorf("exported function names must start with a capital letter")
+	}
+
+	moduleName := tm.t.Module.Name
+	if _, ok := tm.s.moduleCaller.functions[moduleName]; !ok {
+		tm.s.moduleCaller.functions[moduleName] = make(map[string]string)
+	}
+
+	if _, ok := tm.s.moduleCaller.functions[moduleName][name]; ok {
+		return "", fmt.Errorf("function %s in module %s was already exported", name, moduleName)
+	}
+
+	// Right now the function name == template name. This may change in
+	// the future.
+	tm.s.moduleCaller.functions[moduleName][name] = name
+	tm.log.Debug("Exported function", "module.name", moduleName, "function.name", name)
+
+	return "", nil
+}
+
+// Call executes a template function by name with the provided
+// arguments. The function must have been exported by the module that
+// provides it through the [Export] function.
+//
+// Attempting to call a function that does not exist will return an
+// error outside of the first pass where it will return `nil` instead.
+//
+// The template that is called must return a value using the `return`
+// template function, which is only available in this context.
+//
+// In addition, all of the file, stencil and other functions are in the
+// context of the parent template, not the template being called.
+//
+// `.` in a template function acts the same way as it does for
+// [TplStencil.ApplyTemplate] (`stencil.ApplyTemplate`). Meaning, it
+// points to [Values]. The caller passed data is accessible on `.Data`.
+//
+// Example:
+//
+//	// module-a
+//	{{- define "HelloWorld" }}
+//	{{- return (printf "Hello, %s!" .Data) }}
+//	{{- end }}
+//	{{ module.Export "HelloWorld" }}
+//
+//	// module-b
+//	{{ module.Call "module-b.HelloWorld" "Jared" }}
+//	// Output: Hello, Jared
+func (tm *TplModule) Call(name string, args ...any) (any, error) {
+	// Allows args to not be set.
+	if len(args) > 1 {
+		return nil, fmt.Errorf("Call() only takes max two arguments, name and data")
+	}
+
+	// Functions cannot be called in the first pass.
+	if tm.s.isFirstPass {
```

Comment:
this seems liek a bit of a dangeorus road to go down.  now you need 3 passes -- first pass to define functions and basics, second to build library calls with function calls, third to actually do the render.

I wonder if you actually want to repeat passes until state stops changing

Reply:
(and Memoize any native module function calls..?)

Reply:
We've been doing this two pass system for everything already, module hooks, globals, and applytemplate (which is light version of this).

The only trade off there is that you can't dynamically register functions, which is already the case with Go templates anyways. The way this works out is we just return nothing during that first pass where the second pass is the render results. I haven't seen a case where this doesn't work yet.

Reply:
additional context, VK and I talked about the two pass vs render until stop state and both landed on two pass probably working for everything we'll want to support.

Reply:
Will look at changing the state system in the near(ish) future.